onetl/db/db_reader/db_reader.py

```diff
@@ -467,11 +526,22 @@ def validate_options(cls, options, values):
 
         return None
 
-    @validator("hwm_expression", pre=True, always=True)  # noqa: WPS238, WPS231
-    def validate_hwm_expression(cls, hwm_expression, values):
-        connection: BaseDBConnection = values["connection"]
-        dialect = connection.Dialect
-        return dialect.validate_hwm_expression(connection=connection, value=hwm_expression)
+    def detect_hwm(self, hwm: HWM) -> HWM:
+        schema = {field.name.casefold(): field for field in self.get_df_schema()}
+        column = hwm.entity.casefold()
+        target_column_data_type = schema[column].dataType.typeName()
+        hwm_class_for_target = self.connection.Dialect.detect_hwm_class(target_column_data_type)
+        if hwm.value:
+            try:
+                hwm_class_for_target.parse_obj(hwm)
+            except ValueError as e:
+                error_message = (
+                    f"Data type mismatch detected for target column '{column}' - '{target_column_data_type}' "
+                    f"and passed HWM value type: '{type(hwm.value).__name__}'. "
+                    "Please ensure consistency across target and HWM store."
+                )
+                raise ValueError(error_message) from e
+        return hwm_class_for_target.parse_obj(hwm)
```

Comment:
```suggestion
        try:
            return hwm_class_for_target.deserialize(hwm.dict())
        except ValueError as e:
            error_message = textwrap.dedent(
                f"""
                Table column {column!r} has Spark type {target_column_data_type!r}
                which correspond to HWM type {hwm_class_for_target.__name__!r}.
                
                But `DBReader.hwm` object has type {type(hwm).__name__!r}, which is not compatible.
                How to solve this issue:
                * Check that you set correct HWM name (it is unique).
                * Check that HWM store you're using contain valid value for this HWM name.
                * If you set DBReader(hwm=SpecificTypeOfHWM(...)), check it HWM type matches column type.
                * If you set DBReader(hwm=AutoDetectHWM(...)), please check if table schema have not been changed since previous run.
                """,
            )
            raise ValueError(error_message) from e
```
Давай попробуем вот так